### ISSUE TITLE
Negative margin in dimension controls

### ISSUE THREAD

**Issue by carolinan** [WordPress/gutenberg#32644]
## What problem does this address?

It is not possible to use negative margins in the margin control.

## What is your proposed solution?

If negative margins were allowed, designers could overlap blocks easier without having to create custom block styles.
This would also open up for more varied block patterns -the block pattern directory can not use custom block styles.

**Comment by thirumani02** [comment]
Actually you can do it by passing `min` prop. 
for ex. 
```
<BoxControl
    label={__('Margin')}
    onChange={(value) => console.log(value)}
    inputProps={{ min: -300 }}
/>
```

It is not explained in Readme but i can able to find it when searching inside core files

**Comment by timsinc** [comment]
Does anyone know if there are plans to implement this? Negative margins are a very common and useful layout tool that seems should be enabled by default for the core margin control.

**Comment by aurooba** [comment]
Hallo! Trying to work with FSE and this is a serious _serious_ oversight. We need to be able to set negative margins, it's a very common technique.

**Comment by carolinan** [comment]
I am not sure how this would affect the blockgap?

**Comment by mrfoxtalbot** [comment]
+1 to adding negative margins. I just found a case in the wild where this would have been very useful.

**Comment by timsinc** [comment]
> I am not sure how this would affect the blockgap?

I believe a negative margin value would affect the blockgap just like a positive margin does and override the blockgap css for that particular block.

**Comment by daviedR** [comment]
+1

Negative margin is essential for design and layout building

**Comment by samxmunoz** [comment]
+1 

Negative margin allowing for overlapping elements is essential.

**Comment by rolf-yoast** [comment]
+1

I need it for a design I'm currently building.

**Comment by aristath** [comment]
Reopening this one because #40464 is getting reverted.

**Comment by AdsonCicilioti** [comment]
+1 
Negative values for spacing fields. Essential for layout building.

**Comment by jonathanelmgren** [comment]
+1
Negative values for spacing fields. Essential for layout building.

**Comment by peiche** [comment]
+1
Negative values are essential for building layouts.

**Comment by diegosomar** [comment]
+1
Negative values are essential for building layouts (with z-index in the blocks).

**Comment by nendeb** [comment]
+1
Instead of allowing unlimited permits, why not allow up to -20?

**Comment by nendeb** [comment]
+1
Instead of allowing unlimited, how about allowing -20?

**Comment by namithj** [comment]
+1

The is essential to designing layouts, the recent revert broke a few sites of mine.

**Comment by carolinan** [comment]
> +1 Instead of allowing unlimited, how about allowing -20?

This might be possible by using "spacing presets" in theme.json. I have not tested it.

**Comment by nendeb** [comment]
> _This might be possible by using "spacing presets" in theme.json. I have not tested it._

Where can I find the specific code?

**Comment by carolinan** [comment]
https://github.com/WordPress/gutenberg/pull/41527

**Comment by nendeb** [comment]
#41527

Thanks.
This is supposed to define a negative set value.

--wp-priset--spacing-size--minus1: -1
--wp-priset--spacing-size--minus2: -2
--wp-priset--spacing-size--minus3: -3
  .
  .
  .

Before testing.
I don't think this is a very good way to do it if

**Comment by rxnlabs** [comment]
This is definitely needed. I was trying to show a client how to add margins to a block and could not add negative margins when they asked about overlapping a block on top of another. This is a limitation for block themes in being able to achieve certain layouts

**Comment by seothemes** [comment]
@nendeb that will work, however the scale is also applied to padding. We really need support for negative values on margin inputs, it's a huge limitation for designers.

**Comment by nicmare** [comment]
please give the freedom over negative values to the users NOW 👍

**Comment by seothemes** [comment]
There is a workaround to allow negative margins, but without responsive controls they may not be the answer.

Please see https://github.com/WordPress/gutenberg/pull/40464#issuecomment-1260263890

**Comment by bnjunge** [comment]
+1 Negative values

**Comment by dimasoho** [comment]
+1 NEGATIVE MARGINS MUST GO ON!!

**Comment by hagege** [comment]
+1 negative values for margins to realize an outfit like this:
![image](https://user-images.githubusercontent.com/27292550/203539422-5a14b219-750d-4635-9f70-a80775a428e2.png)

**Comment by flipflap** [comment]
I also need it.... please...

**Comment by websitegenii** [comment]
+1 for negative margins
I built a block that sometimes needs to overlap the block below it or above it for the design, yet after implementing I have found that it doesn't support negative. We will have to build a manual workaround which is not ideal.

**Comment by dmpinder** [comment]
Hard to believe we're approaching this issue being 2 years old, and it still isn't implemented yet. It's an essential feature.

**Comment by andrewserong** [comment]
I agree that this would be a very cool feature! It looks like some of the previous attempts stalled for a few reasons (e.g. https://github.com/WordPress/gutenberg/pull/40464). From a quick read over the reasons why allowing negative margins was reverted in the past, it looks like some of the concerns were:

* By allowing setting negative margins, it's easy to accidentally set a block to be either over or underneath another block within the editor canvas (or off screen entirely), making it hard to re-select the block without using the list view
* How should the editor work when selecting blocks that overlap? I.e. if you select a block that is overlapped by another block, while it is selected, should it then be brought forward to be on top via `z-index`?
* If folks are creating offset designs, or designs where blocks deliberately overlap, should the user be able to specify a `z-index` to be able to decide which block is on top? For example, to achieve designs like in the screenshot (https://github.com/WordPress/gutenberg/issues/32644#issuecomment-1324936898) the user needs to be able to decide which block gets to be on top, and which is on the bottom. Implementing `z-index` support could be quite related to position support in that respect (related issue: https://github.com/WordPress/gutenberg/issues/47043)
* Depending on how far the block is offset via negative margins, consideration might need to be given for how the block appears in the list view
* There are also some ongoing conflicts between defining margin rules and the way that layout rules are applied. Some of these issues are being looked into over in https://github.com/WordPress/gutenberg/issues/43404, and it would be good to resolve them before revisiting negative margin support.

For these reasons, I think the feature could probably use some more design work to figure out the ideal UX, as it would likely inform some of the decisions about how to implement negative margins in a way that can be built on reliably, while considering z-index implications for other features (like sticky or fixed headers, or other kinds of positioning).

I'll add the label for "needs design feedback". For context, with my comments here, I'm mostly trying to identify what the practical next steps might be for building out the feature. I very much share everyone's enthusiasm for the feature!

CC: @jameskoster @jasmussen since we've been discussing the position support lately, and I think this would be a very cool feature to consider alongside that work.

**Comment by jameskoster** [comment]
I don't have the full context, so take my (gut instinct) responses with a grain of salt.

> By allowing setting negative margins, it's easy to accidentally set a block to be either over or underneath another block within the editor canvas (or off screen entirely), making it hard to re-select the block without using the list view

I don't see this as an issue. It seems unlikely that you'd configure things so that one block _entirely_ obscures another. In any case, this is one of the main purposes of list view. 

> How should the editor work when selecting blocks that overlap?

I don't see any need to break the wysiwyg?

> If folks are creating offset designs, or designs where blocks deliberately overlap, should the user be able to specify a z-index to be able to decide which block is on top?

Yes, most probably. A z-index control would be quite advanced though, perhaps a shortcut in the ellipsis menu ("Bring to front") could help? I agree there's some overlap with the position controls.

> Depending on how far the block is offset via negative margins, consideration might need to be given for how the block appears in the list view

Would you mind elaborating on that? 

---

A tangential thought, how would a Transform control play here? The ability to offset position (`transform: translateX(-10px)`) would afford the utility of combining positive margins with negative offsets. I'm not sure if that's useful in any practical sense, but perhaps something to keep in mind.

**Comment by AdsonCicilioti** [comment]
It is interesting to think of a concept of dispersion with `transform: translate(<x>,<y>)` and not of margin, but finding a way to deal with the proper visual marking of the selection handles and the toolbar of the object or block .

I want to look at it more calmly to collaborate better.

**Comment by andrewserong** [comment]
Thanks for the responses @jameskoster! I'm mostly trying to distill what appears to be the blockers from earlier attempts, so please take these comments as me attempting to understand why previous attempts were blocked (and what might help unblock the feature in the future) 🙂. Also, I imagine this work wouldn't be until after 6.2, so I don't think there's a particular rush on figuring this out.

> I don't see this as an issue. It seems unlikely that you'd configure things so that one block entirely obscures another.

I think the concern is more about how easy it might be for a user to accidentally get their UI in a state that's difficult to revert from. I.e. if you accidentally set things so that the margin is overlapping or under another block entirely, then it might be difficult to get things back. I think the two things that would help here are the list view (like you mention), and possibly forcing z-index to the top while a block is selected. Without the latter, it could be easy to get stuck.

> I don't see any need to break the wysiwyg?

The problem is that without deliberately setting a z-index, which element sits on top will be slightly random (to the perspective of the user). I think overlapping blocks will require a z-index to be set (either explicitly or implicitly via UI controls). Without this, the wysiwyg approach will match a user's intent some of the time, but not predictably.

> A z-index control would be quite advanced though, perhaps a shortcut in the ellipsis menu ("Bring to front") could help?

Yes, I'd imagine that users likely wouldn't be entering z-index values directly (unless under an advanced controls) and instead there'd be fuzzier UI language for "bring to front", etc 👍

> consideration might need to be given for how the block appears in the list view.... Would you mind elaborating on that?

This might not be as big an issue as I imagine, but depending on the negative values used, is it possible for blocks to be displayed in the editor canvas in an order that's different to how they appear semantically in the list view? Quite possibly not a problem to deal with at the outset, but depending on the site layout, I thought it good to flag that negative margins offer the ability to have a mismatch between the list view and the visual hierarchy within the editor canvas.

---

From my perspective, this would be a terrific feature to land, but since earlier attempts like https://github.com/WordPress/gutenberg/pull/40464 stalled due to digging into some of these issues, I think it might be worth a design exploration to settle on the ideal UX for users building out patterns with overlapping blocks before making another attempt in code. Essentially, figuring out which of the issues are minor annoyances that users can deal with (i.e. if you don't want the block to be off the screen, don't set it to -2000px), and which are potential blockers that would need to have a solution before landing (being able to choose which block sits on top).

Perhaps one way to approach it would be to consider negative margins in the context of relative/absolute positioned blocks — so that it's considered as one part of the whole of dealing with positions for overlapping blocks. That means it could be one of negative margins, transforms, or absolute positioning — or a mix of them all? Whatever the case, I think if we hone in on the question "how should a user create patterns of overlapping blocks" that might help us get closer to an answer.

**Comment by dkjensen** [comment]
+1

**Comment by jasmussen** [comment]
I think I mentioned this codepen on another post, but [using margin-inline we might be able to afford negative margins while still having margin collapsing and the default centered column](https://codepen.io/joen/pen/vYrYqya?editors=1100). It would be a bit of a refactor, but wanted to offer the concept in case it looks feasible.

**Comment by wpalani** [comment]
Definitely a must for building modern layouts. +1

**Comment by andrewserong** [comment]
I've started a draft PR to explore some of the ideas surrounding position controls and control over `z-index`. It's a very early exploratory PR so will likely need a bunch of design work to figure out a viable path forward with decent UX, but hopefully it might help further some of the discussion / help reveal a potential path forward. The exploratory PR is over in #48235.

From playing around with negative margins, one of the reasons why I think we'll need some form of control over `z-index` is that setting a negative margin with the default z-index values doesn't always match up with a desired layout. For example, in the below screenshots, the first one is the desired result when setting a negative top margin for an image to get it to sit over an adjacent block. The second screenshot is the unintended situation where the block sits below the adjacent cover block when setting a negative bottom margin. The third screenshot is with the PR applied, where the block is explicitly set to "Top", which correctly places the negative margin block over the Cover block.

| Negative top margin | Negative bottom margin (auto z-index — the block is covered) | Negative bottom margin (explicit bottom z-index) |
| --- | --- | --- |
| ![image](https://user-images.githubusercontent.com/14988353/220017107-530be5b3-bbf2-4815-bbeb-7485d5b14a9d.png) | ![image](https://user-images.githubusercontent.com/14988353/220015060-49b13f36-213e-405b-801e-d4142fe0b886.png) | ![image](https://user-images.githubusercontent.com/14988353/220015082-9b6628d9-a93e-4ceb-ac14-de04b63db0aa.png) |

I've added a few ideas for next steps to #48235 for continuing to chip away at the problem, but I'd also be very keen to hear if folks have ideas here (or in other issues) about UI/UX approaches for adjusting z-index or desired layer for blocks.

**Comment by nicmare** [comment]
> I've started a draft PR to explore some of the ideas surrounding position controls and control over `z-index`. It's a very early exploratory PR so will likely need a bunch of design work to figure out a viable path forward with decent UX, but hopefully it might help further some of the discussion / help reveal a potential path forward. The exploratory PR is over in #48235.
> 
> From playing around with negative margins, one of the reasons why I think we'll need some form of control over `z-index` is that setting a negative margin with the default z-index values doesn't always match up with a desired layout. For example, in the below screenshots, the first one is the desired result when setting a negative top margin for an image to get it to sit over an adjacent block. The second screenshot is the unintended situation where the block sits below the adjacent cover block when setting a negative bottom margin. The third screenshot is with the PR applied, where the block is explicitly set to "Top", which correctly places the negative margin block over the Cover block.
> 
> Negative top margin	Negative bottom margin (auto z-index — the block is covered)	Negative bottom margin (explicit bottom z-index)
> ![image](https://user-images.githubusercontent.com/14988353/220017107-530be5b3-bbf2-4815-bbeb-7485d5b14a9d.png)	![image](https://user-images.githubusercontent.com/14988353/220015060-49b13f36-213e-405b-801e-d4142fe0b886.png)	![image](https://user-images.githubusercontent.com/14988353/220015082-9b6628d9-a93e-4ceb-ac14-de04b63db0aa.png)
> I've added a few ideas for next steps to #48235 for continuing to chip away at the problem, but I'd also be very keen to hear if folks have ideas here (or in other issues) about UI/UX approaches for adjusting z-index or desired layer for blocks.

really like the simplicity. maybe z-index should be applied automatically when using a negative margin. would save one or two additional clicks. and if its the wrong value, one could change it anyways by hand

**Comment by cawa-93** [comment]
As ugly and unsafe workaround I just open browser inspector and remove `min="0"` from field where I want set negative value. And save. It will store value in HTML markup
![зображення](https://user-images.githubusercontent.com/1662812/225927345-57807b6f-4412-4996-a6df-9ac543f0f97c.png)

**Comment by richtabor** [comment]
> A z-index control would be quite advanced though, perhaps a shortcut in the ellipsis menu ("Bring to front") could help? 

That'd be simple enough. 

> By allowing setting negative margins, it's easy to accidentally set a block to be either over or underneath another block within the editor canvas (or off screen entirely), making it hard to re-select the block without using the list view

What if on select the block was "brought to the front"? 

---

Seems if we add these two bits, we could allow negative values in the margin input fields yea?

**Comment by nicmare** [comment]
> What if on select the block was "brought to the front"?

would not work because of pointer events overlapping elements

**Comment by richtabor** [comment]
> would not work because of pointer events overlapping elements

Could you elaborate please?

**Comment by nicmare** [comment]
> > would not work because of pointer events overlapping elements
> 
> Could you elaborate please?

in many cases a div is not completely filled with content like an image. i.e. if width of image is set 50%. So the containing div seems "transparent" and then blocking pointer events initially targeted to the layer underneath that div. hard to explain. you will see it once you get confronted with it. so there needs to be a setting in navigator to change z-index

**Comment by richtabor** [comment]
Do you mean it'd be hard to select the block behind another block, if it's covered entirely? Preventing the onSelect "bring to front" idea from triggering?

**Comment by richtabor** [comment]
I don't think folks want to put blocks 100% over other blocks with negative margin, but more likely add stylistic offsets. If so, the block further back would need to be selected via the List view. 

Still think it'd be fine to implement. At least get a PR together to test out the limitations.

**Comment by asmittle** [comment]
+1. I use negative margins literally all of the time.

Is this really still not a thing? For instance, I need 0 space between a heading block and an image below it. I already have the block space set to 0, but there's still about 14px of space there. The ability to add -14px of bottom margin to the heading, or -14px of top margin to the image would be ideal in this situation, instead of needing to create a CSS class. Screenshots of comp and results on the page attached.
![mockup](https://github.com/WordPress/gutenberg/assets/100860678/eddc2696-3e48-430e-bab2-1e9d206c8ef2)
![actual-results](https://github.com/WordPress/gutenberg/assets/100860678/04e05cda-a4b3-4414-9cfc-b147acf08386)

**Comment by QuietNoise** [comment]
+1 Just adding my vote and baffled face why it still not part of the core :P <3

And just to not have an empty comment with vote only. This is how I solved it in most of my use cases:
![image](https://github.com/WordPress/gutenberg/assets/3076187/78248c07-07ff-4b40-bde6-b3c9bd5757c2)

I've added a custom control for some blocks just for that feature.  I don't care that it conflicts with core margin settings because it's too important to not having it all, and having it as a separate control with a warning keeps content admins on their toes when using it. Everyone's happy.

**Comment by Pacicio** [comment]
+1 for negative margins, however I found a workaround to make a "fake" negative margin using a gradient background. It's only applicable to some cases but it's better than nothing if you don't want to use custom CSS.

![Fake-negative-margin](https://github.com/WordPress/gutenberg/assets/15251018/c193bc34-360a-4f5d-bae3-03add3a0cd89)

**Comment by nicmare** [comment]
> +1 for negative margins, however I found a workaround to make a "fake" negative margin using a gradient background. It's only applicable to some cases but it's better than nothing if you don't want to use custom CSS.
> 
> ![Fake-negative-margin](https://user-images.githubusercontent.com/15251018/250580440-c193bc34-360a-4f5d-bae3-03add3a0cd89.gif)

That’s how I do it. You could also combine it with the 2nd Color 0% opacity and a cover block (nested) to create interesting effects with images. But there is sometime a glitch of 1 pixel.

**Comment by adamghorne** [comment]
Plus one for this.

**Comment by itsjustdj** [comment]
Plus one for this issue. I needed to use negative margins on a recent client project and had to revert to customize.php to add custom CSS and make it work. It would be nice if you could just enter a negative margin in the site editor.

**Comment by metabreakr** [comment]
+1

**Comment by ohiawins** [comment]
+1 soooohoohooho many people are gonna be totally stoked when this comes out.

**Comment by thomasnuggets** [comment]
I write a snippets to put in your "custm snippets" plugin : 

```php
add_action('admin_footer', 'remove_min_value_inline_script');

function remove_min_value_inline_script() {
    echo '<script type="text/javascript">
        document.addEventListener("DOMContentLoaded", function() {
            const observer = new MutationObserver(function(mutations) {
                mutations.forEach(function(mutation) {
                    if (mutation.addedNodes.length) {
                        mutation.addedNodes.forEach(function(node) {
                            if (node.querySelectorAll) {
                                node.querySelectorAll("input[min=\'0\']").forEach(function(input) {
                                    input.removeAttribute("min");
                                });
                            }
                        });
                    }
                });
            });

            observer.observe(document.body, {
                childList: true,
                subtree: true
            });
        });
    </script>';
}
``` 

And It's work perfectly :)

**Comment by nicmare** [comment]
> I write a snippets to put in your "custm snippets" plugin :
> 
> ```
> add_action('admin_footer', 'remove_min_value_inline_script');
> 
> function remove_min_value_inline_script() {
>     echo '<script type="text/javascript">
>         document.addEventListener("DOMContentLoaded", function() {
>             const observer = new MutationObserver(function(mutations) {
>                 mutations.forEach(function(mutation) {
>                     if (mutation.addedNodes.length) {
>                         mutation.addedNodes.forEach(function(node) {
>                             if (node.querySelectorAll) {
>                                 node.querySelectorAll("input[min=\'0\']").forEach(function(input) {
>                                     input.removeAttribute("min");
>                                 });
>                             }
>                         });
>                     }
>                 });
>             });
> 
>             observer.observe(document.body, {
>                 childList: true,
>                 subtree: true
>             });
>         });
>     </script>';
> }
> ```
> 
> And It's work perfectly :)

thats a very very generic code. i would at least add this to the selector:
`querySelectorAll(".spacing-sizes-control input[min=\'0\']")`

**Comment by thomasnuggets** [comment]
> 

I adapted the code : 
```php
add_action('admin_footer', 'remove_min_value_inline_script');

function remove_min_value_inline_script() {
    echo '<script type="text/javascript">
        document.addEventListener("DOMContentLoaded", function() {
            const observer = new MutationObserver(function(mutations) {
                mutations.forEach(function(mutation) {
                    if (mutation.addedNodes.length) {
                        mutation.addedNodes.forEach(function(node) {
                            if (node.querySelectorAll) {
                                node.querySelectorAll(".spacing-sizes-control input[min=\'0\']").forEach(function(input) {
                                    input.removeAttribute("min");
                                });
                            }
                        });
                    }
                });
            });

            observer.observe(document.body, {
                childList: true,
                subtree: true
            });
        });
    </script>';
}

```

**Comment by wiinf** [comment]
+1 

It would realy be very helpful!!

**Comment by alexisreau** [comment]
+1
Valeurs négatives pour les champs d'espacement. Indispensable pour la construction d’agencements.

**Comment by JiveDig** [comment]
+1

Adding a note that fixed (px, em, etc.) values shouldn't be the only option. I'd like to see the spacing scale allow for negative clamp/min/max values as well. 

From [Andrews screenshots](https://github.com/WordPress/gutenberg/issues/32644#issuecomment-1436361368) you can see how the padding on the red Group block could have a value that uses clamp from the spacing scale. Adding a set px/em value for negative margin will not scale accordingly along with the Group and very likely cause unintended overlap issues.

We've used custom margin settings for a while (before they were in core) with this approach and it's proven to work fairly well.

**Comment by ouw-jvt** [comment]
Big +1 to this.

Reviewing a client design right now and I really could use this functionality. I will probably wind up home-baking something

**Comment by madfcat** [comment]
We need this asap! Why it is not implemented by default though?

**Comment by curiousduck-dev** [comment]
+1! 

Is there an update on this for 2024?

**Comment by flabernardez** [comment]
Need it! Please! Now I'm adding it with a css class 😔

**Comment by alperta** [comment]
+1 so that we don't need to write custom css classes.

**Comment by wpsoul** [comment]
OMG, it still doesn't have negative margin?

**Comment by DevAndreyBilchenko** [comment]
+1 topic started from 2021

**Comment by SH4LIN** [comment]
+1

**Comment by lunule** [comment]
+1

**Comment by cloudcreationsjess** [comment]
+1

**Comment by annezazu** [comment]
As this is being explored in this initial PR (https://github.com/WordPress/gutenberg/pull/60347#), I want to share some UX considerations. Specifically, I want to note that there needs to be a pause/stop point before going to negative values when using either the slider or the ability to drag up/down to change controls as shown below:

https://github.com/WordPress/gutenberg/assets/26996883/0cdb1c95-fa1b-424f-a7fc-f386e3c70734

Right now, with the current PR that is meant to be a small start towards solving this, you can quickly slide past the 0 and it would go a long way to have a "sticky" point at 0 as it's a common action to try to 0 out values to reset something. I'm aware, I can go to the three dot menu to reset but this would keep the more visible options useful to reset while also allowing new functionality.
